### PR TITLE
retry when epoll_wait fails with EINTR

### LINKED ISSUE #971: UnhandledException in WasEventDetected method
On my test system the method _WasEventDetected_ from the _SysFsDriver.Linux.cs_ throws an Exception because the _Interop.epoll_wait_ method fails. While I haven't found the reason yet, how am I supposed to catch this exception? Since this code is executed in a new internal Thread, it will jump to my unhandled exception handler and kills my application.

### PR DESCRIPTION
**Pull request by tibel**
Fixes #971

Implement retry handling for `epoll_wait` in `SysFsDriver.WasEventDetected` similar to `LibGpiodDriverEventHandler` implementation.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/iot/pull/1801)

### CODE REVIEW

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/SysFsDriver.cs
Comment (pgrawehr) [inline src/System.Device.Gpio/System/Device/Gpio/Drivers/SysFsDriver.cs:470]: Please include the error code in the message.

**Review by pgrawehr — APPROVED** [review]
LGTM

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/SysFsDriver.cs
Comment (tibel) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/SysFsDriver.cs:470]: @pgrawehr Do you mean the comment or exception message?

I didn't want to change the exception message in this pull request.
What about using `ExceptionHelper` for the exception (also on other places in `SysFsDriver`)?
Personally I prefer to handle that in a separate PR to keep this one simple.

**Comment by tibel** [comment]
Any chance this gets merged and released soon?

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/SysFsDriver.cs
Comment (krwq) [inline src/System.Device.Gpio/System/Device/Gpio/Drivers/SysFsDriver.cs:463]: we also have `epoll_wait` on line 448 in the same file, would be good to also retry there

**Review by krwq — APPROVED** [review]
LGTM after fixing the other instance

**Comment by krwq** [comment]
@tibel we will bring the subject of creating minor release during our next triage meeting. My rough guess we will want at minimum https://github.com/dotnet/iot/pull/1789 merged and will likely wait for this one to be merged as well. Both PRs are really close to be merged.

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/SysFsDriver.cs
Comment (tibel) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/SysFsDriver.cs:463]: I was not sure how to handle that one as the result is ignored.
```c#
// Ignore first time because it will always return the current state.
Interop.epoll_wait(pollFileDescriptor, out _, 1, 0);
```

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/SysFsDriver.cs
Comment (krwq) [inline src/System.Device.Gpio/System/Device/Gpio/Drivers/SysFsDriver.cs:470]: might be a good idea to add error code in this message as well for easier debugging

Comment (tibel) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/SysFsDriver.cs:470]: My idea is to change it to use `ExceptionHelper` in a follow-up PR.

Comment (krwq) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/SysFsDriver.cs:470]: I see @pgrawehr already asked for that. I'd prefer we changed it in this PR, it's not much different concern. I'd just add append something like ` (ErrorCode={ERROR_CODE_EINTR})` and change string to interpolated. We can create another issue to have more uniform way of throwing but it's better to improve it now as we're touching this code

Comment (krwq) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/SysFsDriver.cs:470]: if you're planning to send a follow up then that's fine too but just want to make sure we actually do that - usually it's simpler to just do it in a single PR, also we can do the temporary fix now and follow up better fix in case you forget to send follow-up 😄 (sorry but happens to everyone for various reasons)

Comment (tibel) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/SysFsDriver.cs]: Added error code to message as requested.

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/SysFsDriver.cs
Comment (krwq) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/SysFsDriver.cs:463]: add variable and while loop, just retry on EINTR but ignore errors as we're not sure how to handle them in there

**Review by krwq — APPROVED** [review]

**Review by joperezr — APPROVED** [review]

**Comment by joperezr** [comment]
Thanks for the fix @tibel we will discuss tomorrow about our next minor release during our triage meeting and will let you know when that will be so you can expect news about that tomorrow.

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/SysFsDriver.cs
Comment (krwq) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/SysFsDriver.cs:471]: honestly I'm not sure so good to have just single thing fixed in here.

**Pull request merged**
